### PR TITLE
Fix llava half precision and autocast issues

### PR DESCRIPTION
**Pull request by frasermince**
What does this PR do?

Currently half precision training on LLaVa is broken due to the language embedding output always being full precision. This PR matches the behavior of the original LLaVa implementation by casting the result of language embedding to the image_features' dtype. See [here](https://github.com/haotian-liu/LLaVA/blob/main/llava/model/multimodal_encoder/clip_encoder.py#L51) for relevant line in the original repo.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Could you add a test e.g. like `create_and_check_reformer_model_fp16_forward` that we have for other models?

**Comment by frasermince** [comment]
> Thanks for adding this!
> 
> Could you add a test e.g. like `create_and_check_reformer_model_fp16_forward` that we have for other models?

Apologies on the wait on this. After implementing the test, it looks like the problem here is not quite what I thought it was. I will try to respond soon with either fixing the correct problem or closing this PR.

**Comment by frasermince** [comment]
After some further investigation, I tracked down exactly what was happening. `.half()` or `.to(torch.bfloat16)` work fine on the model. However, the issue arises when the model is run within `with torch.autocast`. This also includes if you are using a trainer with the `bf16` or `fp16` arg set to true since `accelerate` uses an `autocast` context behind the scenes. When run in autocast bf16 or fp16 block the [language model embedding](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L419) has full precision. However, the [`image_features`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L436) are in half precision. Due to these being merged there is a type error. To fix this we need to either cast embeddings to half precision or image features to full. It made the most sense to me to cast the embeddings to half precision. Let me know if you prefer we go the other direction but this seemed the most efficient.

I also fixed the same issue on llava_next. In the process I discovered that there were NaNs occuring upon running llava_next in half precision. To fix this I used `randn` instead of `empty` similarly to the [original llava initialization](https://github.com/haotian-liu/LLaVA/blob/4e2277a060da264c4f21b364c867cc622c945874/llava/model/llava_arch.py#L85).

File: src/transformers/models/llava_next/modeling_llava_next.py [hunk 1]
```diff
@@ -306,8 +306,9 @@ def __init__(self, config: LlavaNextConfig):
         self.vision_tower = AutoModel.from_config(config.vision_config)
 
         self.multi_modal_projector = LlavaNextMultiModalProjector(config)
-
-        self.image_newline = nn.Parameter(torch.empty(config.text_config.hidden_size, dtype=self.dtype))
+        self.image_newline = nn.Parameter(
+            torch.randn(config.text_config.hidden_size, dtype=config.torch_dtype or self.dtype)
```
Comment (frasermince) [inline src/transformers/models/llava_next/modeling_llava_next.py:310]: We needed some way to set this half precision in autocast settings. Let me know if this is incorrect. I am still unclear when to use `torch_dtype` and when not to.

File: tests/models/llava/test_modeling_llava.py [hunk 2]
```diff
@@ -158,6 +159,19 @@ def prepare_config_and_inputs_for_common(self):
         }
         return config, inputs_dict
 
+    def create_and_check_llava_model_fp16_forward(self, config, input_ids, pixel_values, attention_mask):
+        model = LlavaForConditionalGeneration(config=config)
+        model.to(torch_device)
+        model.eval()
+        with torch.autocast(device_type="cuda", dtype=torch.float16):
```
Comment (frasermince) [inline tests/models/llava/test_modeling_llava.py:166]: I had some concerns about putting this here given I haven't seen autocasting tested in the unit tests elsewhere in the repo. Let me know if you prefer this being tested through an integration test with accelerate.

File: src/transformers/models/llava_next/modeling_llava_next.py (hunk 1 above)
Comment (frasermince) [reply src/transformers/models/llava_next/modeling_llava_next.py:310]: Also I should note I am not quite certain `randn` is correct here but the `empty` was causing NaNs.

Comment (amyeroberts) [reply src/transformers/models/llava_next/modeling_llava_next.py:310]: Just from looking at this, I suspect the issue is that the param `self.image_newline` is being create as an empty tensor, but it's not then being initialized in `_init_weights`

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this!

File: tests/models/llava/test_modeling_llava.py (hunk 2 above)
Comment (frasermince) [reply tests/models/llava/test_modeling_llava.py:166]: It's kind of the opposite. It works fine without autocasting in `.half`. The issue is when using a trainer with the `fp16` or `bf16` flag the model returns a type error. This uses autocasting behind the scenes through accelerate so I wrote these tests as the simplest case to capture this failure. I was not quite sure how to handle this in these tests given we are not testing autocasting behavior elsewhere.

Comment (frasermince) [reply tests/models/llava/test_modeling_llava.py:166]: Correct, the issue arises in passing the model to the trainer with the `fp16` or `bf16` flags. The test you included would work just fine. As far as I can tell this is due to how the model works when autocast is used within the trainer (indirectly through accelerate). I was able to replicate the same bug in this test using the autocast block as I see when I run the model through the trainer.

Comment (amyeroberts) [reply tests/models/llava/test_modeling_llava.py:166]: If this PR is to address the trainer issue you mention, then I think we might want to open a new github issue, where this can be tracked and discussed. The issue with the fp16 and bf16 flags might be a larger one, independent of llava

Comment (frasermince) [reply tests/models/llava/test_modeling_llava.py:166]: Fair enough. Based on looking at the trainer code it does not appear the trainer does anything incorrectly. It's the typing of Llava in particular that is incorrect in these autocast settings. Agreed however this could apply to other models as well.

Comment (amyeroberts) [reply tests/models/llava/test_modeling_llava.py:166]: Ah, OK. I think we should do this in two steps. Adding an integration test for trainer & autocast for models sounds like a good idea. I suspect it might throw up quite a few things to address and the design of the test is important to make sure it's as lightweight as possible. By splitting up, we can add this fix in quickly and then iterate on the test design / fixing errors it throws.

What I would suggest is:
* Keep the change to llava next and the `create_and_check_llava_next_model_fp16_forward` test in this PR
* Open a new PR for adding an integration test. This would possibly sit under `tests/trainer/test_trainer.py` - @muellerzr will be able to advise here re where and design :)

**Comment by muellerzr** [comment]
I’ll need to dig into this when I’m back to my computer, but @amyeroberts just to describe how the autocast system works in accelerate quickly:

We wrap around the models’ .forward() with an autocast manager, so that the original weights remain in fp32. Not doing so leads to a large deal of precision and end accuracy issues, which is why we maintain this reference. It uses a pinch more memory (than pure doing `model.half()` but for good reason. 

If certain weights need to be cast for certain models, I think that can be okay and taken as a case-by-case basis, but I want to re-read this discussion carefully to understand why.

File: tests/models/llava/test_modeling_llava.py (hunk 2 above)
Comment (muellerzr) [reply tests/models/llava/test_modeling_llava.py:166]: Fully agree with this way forward @frasermince !

Comment (muellerzr) [reply tests/models/llava/test_modeling_llava.py:166]: Let's also document the test with some comments explaining why it's needed

Comment (frasermince) [reply tests/models/llava/test_modeling_llava.py:166]: Sounds like a plan! I'll get to work on that! Would we want this new integration test still namespaced under the model something like tests/llava/trainer/test_trainer.py or are you suggesting we add testing more generally to the trainer?

Comment (muellerzr) [reply tests/models/llava/test_modeling_llava.py:166]: Maybe `test/modeling/llava/test_trainer_llava.py`

since for now its model specific?

**Review by muellerzr — APPROVED** [review]
Overall I think these are good tests how they are, let's please keep a close eye on it and see what other models have issues. We may want to change some things with Accelerate (like specifying certain layers to be in the precision fully) if tests show no loss in performance :)

**Comment by frasermince** [comment]
> Overall I think these are good tests how they are, let's please keep a close eye on it and see what other models have issues. We may want to change some things with Accelerate (like specifying certain layers to be in the precision fully) if tests show no loss in performance :)

Sounds good! I will also add the trainer test in a separate PR as requested.

**Comment by frasermince** [comment]
@muellerzr before merging do you have any suggestions on ensuring accuracies remain as expected? Is it basically taking the assertions in some of the integration tests after a couple of training steps?

File: src/transformers/models/llava_next/modeling_llava_next.py (hunk 1 above)
Comment (frasermince) [reply src/transformers/models/llava_next/modeling_llava_next.py:310]: Right I will move this initialization to _init_weights

**Comment by muellerzr** [comment]
>  Is it basically taking the assertions in some of the integration tests after a couple of training steps?

Yes indeed!

**Review by muellerzr — APPROVED** [review]

File: src/transformers/models/llava_next/modeling_llava_next.py (hunk 1 above)
Comment (frasermince) [reply src/transformers/models/llava_next/modeling_llava_next.py]: At a second glance it looks sufficient to do it on initialization. The original repo just had a a separate function to initialize the vision tower.

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing this and iterating with us to find a solution! 

Only last thing is to run the slow model tests: 

```
RUN_SLOW=1 pytest tests/models/llava tests/models/llava_next
```

Once they're all confirmed to be passing we're good to merge!

**Review by amyeroberts — APPROVED** [review]

**Comment by frasermince** [comment]
> ```
> RUN_SLOW=1 pytest tests/models/llava tests/models/llava_next
> ```

There was one that seemed to be already failing before my changes. I took the liberty of changing the assertion slightly. Let me know if this is acceptable! There are a couple failing in main for me for llava_next as well but that would require me changing the expected logits. Not sure if that is acceptable or not.

**Comment by muellerzr** [comment]
@amyeroberts leave to your discretion :)

**Comment by frasermince** [comment]
Given slow the test was failing on main I changed the logits in the test for now. If that is wrong I can remove.

**Comment by amyeroberts** [comment]
@frasermince Great - thanks for checking! I'm going to be a bit annoying here - as the differences are so small it's quite possible they're just coming from running on different hardware / slightly different env setups. So we might need to revert. 

@ydshieh How can we run the slow tests on this branch in the new github actions workflow to make sure the values are aligned?

**Comment by ydshieh** [comment]
@amyeroberts I am on it and should be ready soon (even possible today 🔥 )

**Comment by ydshieh** [comment]
It is in #30540 :-)

**Comment by amyeroberts** [comment]
Thanks for triggering the tests! OK, so we're hitting OOM errors, which is going to be a pain to try and resolve as it's a wider issue. What I suggest is:
* We resolve the conflicts
* We merge this in 
* If there's some issues with the slow tests which aren't oom, I'll handle in a follow up PR (cc @ydshieh - I'll be the person to ping)

**Comment by amyeroberts** [comment]
@frasermince Thanks for all of your patience iterating with us. We're good to merge!
